rust/common-wasi/src/stdio.rs

```diff
@@ -0,0 +1,598 @@
+pub struct AsyncStdinStream(Arc<Mutex<crate::pipe::AsyncReadStream>>);
+
+impl AsyncStdinStream {
+    pub fn new(s: crate::pipe::AsyncReadStream) -> Self {
+        Self(Arc::new(Mutex::new(s)))
+    }
+}
+
+impl StdinStream for AsyncStdinStream {
+    fn stream(&self) -> Box<dyn HostInputStream> {
+        Box::new(Self(self.0.clone()))
+    }
+    fn isatty(&self) -> bool {
+        false
+    }
+}
+
+impl HostInputStream for AsyncStdinStream {
+    fn read(&mut self, size: usize) -> Result<bytes::Bytes, StreamError> {
+        match self.0.try_lock() {
+            Ok(mut stream) => stream.read(size),
+            Err(_) => Err(StreamError::trap("concurrent reads are not supported")),
+        }
+    }
+    fn skip(&mut self, size: usize) -> Result<usize, StreamError> {
+        match self.0.try_lock() {
+            Ok(mut stream) => stream.skip(size),
+            Err(_) => Err(StreamError::trap("concurrent skips are not supported")),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Subscribe for AsyncStdinStream {
+    async fn ready(&mut self) {
+        self.0.lock().await.ready().await
+    }
+}
+
+mod worker_thread_stdin;
+pub use self::worker_thread_stdin::{stdin, Stdin};
+
+/// Similar to [`StdinStream`], except for output.
+pub trait StdoutStream: Send {
+    /// Returns a fresh new stream which can write to this output stream.
+    ///
+    /// Note that all output streams should output to the same logical source.
+    /// This means that it's possible for each independent stream to acquire a
+    /// separate "permit" to write and then act on that permit. Note that
+    /// additionally at this time once a permit is "acquired" there's no way to
+    /// release it, for example you can wait for readiness and then never
+    /// actually write in WASI. This means that acquisition of a permit for one
+    /// stream cannot discount the size of a permit another stream could
+    /// obtain.
+    ///
+    /// Implementations must be able to handle this
+    fn stream(&self) -> Box<dyn HostOutputStream>;
+
+    /// Returns whether this stream is backed by a TTY.
+    fn isatty(&self) -> bool;
+}
+
+impl StdoutStream for pipe::MemoryOutputPipe {
+    fn stream(&self) -> Box<dyn HostOutputStream> {
+        Box::new(self.clone())
+    }
+
+    fn isatty(&self) -> bool {
+        false
+    }
+}
+
+impl StdoutStream for pipe::SinkOutputStream {
+    fn stream(&self) -> Box<dyn HostOutputStream> {
+        Box::new(*self)
+    }
+
+    fn isatty(&self) -> bool {
+        false
+    }
+}
+
+impl StdoutStream for pipe::ClosedOutputStream {
+    fn stream(&self) -> Box<dyn HostOutputStream> {
+        Box::new(*self)
+    }
+
+    fn isatty(&self) -> bool {
+        false
+    }
+}
+
+/// This implementation will yield output streams that block on writes, and
+/// output directly to a file. If truly async output is required, [`AsyncStdoutStream`]
+/// should be used instead.
+pub struct OutputFile {
```

Comment:
There are several references to stdlib's `Stdout` and `Stderr`. Remove them. Also `OutputFile` and other file-based streams may be removed safely. Pretty much any reference to `std::io` here.